### DISCUSSION TITLE
Access to REOLINK APIs

### DISCUSSION THREAD

**Discussion by ptyers** [fwestenberg/reolink_dev#392] [General]
Reolink (implied) company policy appears to be that access to documentation on the camera APIs is company confidential.

As will be seen in the following email exchange there might be a small chink in that armour, but it will involve a joint effort to persuade a company suit that Open Source Development of applications to service the installed base (current and future) of Reolink cameras on Home Assistant is a volume market.

**Is the effort worth it??????**

EMAIL EXCHANGE BETWEEM MYSELF AND REOLINK TECH SUPPORT 

ME -------------------------------------
The various cameras can support HTTP POST "{"cmd": "GetMdState", "param": {"channel": 0 }}
is there an equivalent which allows remote setting to indicate motion has been seen?
One can SetAudioAlarm or depending on AlarmVersion SetAudioAlarmV20 but the siren on DUO 
and other cameras will not activate until motion is seen.
Given that the Reolink App can, as part of set audio, cause the siren to activate there must be a means of doing this without having motion seen.
My copy of "Camera HTTP API User Guide" Version 4 2019-9 lists no API's related to audio settings, can you please supply information on these.

REOLINK ------------------------------------------------
Thank you for contacting Reolink. 
Sorry that we do not have tests on embedding camera streaming on public websites so far, so I am afraid the support regarding this issue we can provide is limited. But we have attached a PDF document in this email including all available commands of our cameras, you may like to check if it may be of some use to you. Thanks! 
Best Regards
Reolink Support Team – Justina
Attachment(s)
 a link to
 https://support.reolink.com/attachments/token/Ug3ysKYCgE8dxw5SLVBmR0Snc/?name=Reolink-url-command-v2.docx

ME --------------------------------------------
I am totally disbelieving that the provided document covers all available API calls given that I have in my hand a 391 page document doing just that, except it is only a 2019 version. I must assume since audio capabilities were recently added that an updated version of it exists.

REOLINK --------------------------------
I am L2, and I checked with our engineer team.
Right now the API command we can provide is limited.

The doc we attached in last email is the new version we updated this year.
Right now it already include get motion detection URL, but it's local motion detection instead of remote settings.

All other commands are not shared out right now. 

ME ----------------------------------------------

Thank you for your  response, even if it was disappointing.

If I interpret correctly what you are saying then I must deduce the following

1. Despite the fact that Reolink has achieved a significant penetration into the 100,000+ Home Assistant (Open Source home automation application) community where there's a 47% usage of the FFMPEG integration, 6.7% usage of the generic IP  camera integration and a further 3.9% usage of the ONVIF integration. Of this usage there is a significant but not currently tracked penetration by Reolink. I do not have at hand the statistics for the usage of the specific Reolink integration add-on that my colleagues and I have developed but we believe it to be significant. Your response indicates that Reolink is not desiring to improve that penetration.

2. Are you formally stating that Reolink considers documentation of its Application Program Interface (API) as commercial-in-confidence?

3. Are you stating that the 391 page document I refer to is not produced by Reolink?

4. Are you saying that if we wish to enhance usage of Reolink product within that 100,000+ installed base we must reverse engineer your apps to determine how to add necessary features that the camera is capable of providing, such as usage of the siren capability as part of a panic alarm?

REWOLINK ---------------------------------------------

Dear Peter Tyers,          

Sorry for the trouble caused to you.

May i know whether you have any project or plan that need many Reolink cameras, 
and this project need Reolink to provide the API document?
If so, we can check again with sales team, and they can contact you about details.

If just personal use, I am afraid that the Application Program Interface (API) are not available or shared out. 

ME ---------------------------------------------------------

Justina
Thank you for your response
It did not answer my queries but I now realise that asking you to formally state company 
policy was probably unfair.
Your suggestion is good and I will consult with my co-collaborators in France and Holland about making a formal proposal to a person with suitable authority to determine policy regarding distribution of APIs on an international basis.
Perhaps you can suggest an appropriate contact within Reolink International to whom such a proposal could be made?

AS a technical specialist you will be aware of how Open Source Application Development occurs so you know there is no formal company structure which can be referred to.

If you are referring this up the chain of command may I point out that of that 100,000+ installed base of Home Assistant some 47% use the FFMPEG integration which is necessary if camera video streams are being used. How many of these cameras are Reolink I don’t know.

If you can provide that contact point then, for the moment, it would appear that further action on this matter at your level is unnecessary and you may close the case.

Peter Tyers

**Comment by cpainchaud** [comment]
This is interesting, at the same time Reolink admits in a places that they're understaffed which explains why there are so many bugs.

Like this reddit I started a few weeks ago: https://www.reddit.com/r/reolinkcam/comments/qzkhwk/about_api_versions_and_their_respect_on_each_model/

While their API is easy to reverse engineer, we need them to fix the many bugs they have and also rebase their code so each camera doesn't get an instance of the code which has a separate cycle and require bugs to be fixed in every single tree.

**Comment by ptyers** [comment]
Fwestenberg and cpainchaud
Is there any available statistics as to how many users of the HACS Reolink add-on exist. If we are to make an approach to get APIs available we need all the ammunition we can gather. Numbers of forks and pull requests are not likely to be reliable of course.

**Reply by cpainchaud** [reply]
It's installed via HACS so I don't know if HA collects it

**Comment by ptyers** [comment]
INTERESTING

exchange on reddit which I entered into
NOTE THE LAST COMMENT

User avatar
level 1
ricardo-chen99
·
6 days ago
**Reolink Tech**
The command function is normal. To use this command, you need to carry json data. You can't access it with url alone. You can send the command through the Google plug-in“advanced rest client”.

https://imgur.com/a/3irnnQW


xspecial1992
OP
·
ricardo-chen99
·
5 days ago
**Reolink Tech**
just wait for a few days

xspecial1992
OP
·
5 days ago
Ok, thanks!

peter_t1
·
2 hr. ago
You presented info re AudioAlarmPlay API in this forum. I presume it came from an updated version of "Camera HTTP API User Guide" is the latest version available somewhere on the net? My copy is 2019-9 version


Willson1_
·
28 min. ago
**Reolink Admin** 
We are improving the document. I will post the document once we finish.

**Reply by ptyers** [reply]
As of 1700 AEST Wednesday current promise is distribution of updated documentation next week - reddit

**Reply by ptyers** [reply]
A thought. When the updated document becomes available, would it be worth adding to the reolink (not reolink_dev) repo?
Would save contributor's from having to battle tech support for answers which should be readily available.

**Reply by cpainchaud** [reply]
we will see the quality first. Like I said even internally they don't have the same opinion about an API call or another so ....

**Reply by ptyers** [reply]
Here's a copy. Is it worth putting into the repo? Possibly convert to pdf or leave as word?

[Camera HTTP API.zip](https://github.com/fwestenberg/reolink_dev/files/7696429/Camera.HTTP.API.zip)
